### ISSUE TITLE
Ramen end-to-end script fails to deploy open-cluster-management-hub components.

### ISSUE THREAD

**Issue by vdeenadh** [RamenDR/ramen#407]
Invoking the Ramen end-to-end script on a bare metal machine fails to deploy `open-cluster-management-hub` components.

```
# Invoke the end-to-end script
$ hack/ocm-minikube-ramen.sh
:
+ kubectl --context hub apply -k /tmp/veera/open-cluster-management/registration/cluster-manager/hub
clustermanager.operator.open-cluster-management.io/cluster-manager created
+ date
Wed Mar 16 21:05:51 PDT 2022
+ kubectl --context hub -n open-cluster-management wait deployments/cluster-manager --for condition=available
deployment.apps/cluster-manager condition met
+ date
Wed Mar 16 21:06:05 PDT 2022
+ until_true_or_n 90 kubectl --context hub -n open-cluster-management-hub wait deployments --all --for condition=available --timeout 0
Wed Mar 16 21:06:05 PDT 2022
error: no matching resources found
error: no matching resources found
:
+ kubectl --context hub -n open-cluster-management-hub wait deployments --all --for condition=available --timeout 0
error: no matching resources found
+ set -- 1
+ trap - EXIT

# Listing namespaces on the hub shows that open-cluster-management-hub namespace doesn't exist.
$ kubectl --context hub get ns                                  
NAME                      STATUS   AGE
default                   Active   39m
kube-node-lease           Active   39m
kube-public               Active   39m
kube-system               Active   39m
open-cluster-management   Active   39m

# But the cluster-manager deployment is ready in the open-cluster-management namespace
$ kubectl --context hub get deploy -nopen-cluster-management
NAME              READY   UP-TO-DATE   AVAILABLE   AGE
cluster-manager   1/1     1            1           9m25s

# However, cluster manager runs into errors shown below. 
$ kubectl logs --context hub  -nopen-cluster-management deploy/cluster-manager
W0317 04:06:04.009450       1 cmd.go:213] Using insecure, self-signed certificates
I0317 04:06:04.236035       1 observer_polling.go:159] Starting file observer
W0317 04:06:04.262826       1 builder.go:230] unable to get owner reference (falling back to namespace): pods is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot list resource "pods" in API group "" in the namespace "open-cluster-management"
I0317 04:06:04.263793       1 builder.go:262] clustermanager version v0.0.0-unknown-7aa41cd
W0317 04:06:04.744866       1 secure_serving.go:69] Use of insecure cipher 'TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256' detected.
W0317 04:06:04.744891       1 secure_serving.go:69] Use of insecure cipher 'TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256' detected.
W0317 04:06:04.745409       1 builder.go:321] unable to get cluster infrastructure status, using HA cluster values for leader election: infrastructures.config.openshift.io "cluster" is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot get resource "infrastructures" in API group "config.openshift.io" at the cluster scope
I0317 04:06:04.745820       1 leaderelection.go:248] attempting to acquire leader lease open-cluster-management/clustermanager-lock...
I0317 04:06:04.745963       1 event.go:285] Event(v1.ObjectReference{Kind:"Namespace", Namespace:"open-cluster-management", Name:"open-cluster-management", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'ClusterInfrastructureStatus' unable to get cluster infrastructure status, using HA cluster values for leader election: infrastructures.config.openshift.io "cluster" is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot get resource "infrastructures" in API group "config.openshift.io" at the cluster scope
I0317 04:06:04.807824       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I0317 04:06:04.807902       1 shared_informer.go:240] Waiting for caches to sync for RequestHeaderAuthRequestController
I0317 04:06:04.808151       1 configmap_cafile_content.go:201] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I0317 04:06:04.808189       1 shared_informer.go:240] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0317 04:06:04.808619       1 configmap_cafile_content.go:201] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I0317 04:06:04.808805       1 shared_informer.go:240] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0317 04:06:04.809423       1 dynamic_serving_content.go:131] "Starting controller" name="serving-cert::/tmp/serving-cert-3282788192/tls.crt::/tmp/serving-cert-3282788192/tls.key"
I0317 04:06:04.810881       1 secure_serving.go:266] Serving securely on [::]:8443
I0317 04:06:04.811390       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
E0317 04:06:04.814462       1 leaderelection.go:334] error initially creating leader election record: leases.coordination.k8s.io is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot create resource "leases" in API group "coordination.k8s.io" in the namespace "open-cluster-management"
I0317 04:06:04.909062       1 shared_informer.go:247] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0317 04:06:04.909247       1 shared_informer.go:247] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0317 04:06:04.909354       1 shared_informer.go:247] Caches are synced for RequestHeaderAuthRequestController
E0317 04:06:46.367604       1 leaderelection.go:330] error retrieving resource lock open-cluster-management/clustermanager-lock: leases.coordination.k8s.io "clustermanager-lock" is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "open-cluster-management"
E0317 04:07:35.887935       1 leaderelection.go:330] error retrieving resource lock open-cluster-management/clustermanager-lock: leases.coordination.k8s.io "clustermanager-lock" is forbidden: User "system:serviceaccount:open-cluster-management:cluster-manager" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "open-cluster-management"

# I'm not sure if these errors have anything to do with the failure to deploy open-cluster-management-hub components.
```

**Comment by hatfieldbrian** [comment]
I'm addressing this by using the registration-operator image that matches the manifests.  Please retest with fix that closes this issue.

**Comment by hatfieldbrian** [comment]
I reproduced this error using `latest` registration operator image, deleted the `cluster-manager` deployment, then redeployed it  using the image matching the manifests and verified that the `open-cluster-management-hub deployments` were all `available`:
```sh
+ until_true_or_n 90 kubectl --context hub -n open-cluster-management-hub wait deployments --all --for condition=available --timeout 0
Thu Mar 17 16:35:23 PDT 2022
...
deployment.apps/cluster-manager-placement-controller condition met
deployment.apps/cluster-manager-registration-controller condition met
deployment.apps/cluster-manager-registration-webhook condition met
deployment.apps/cluster-manager-work-webhook condition met
Thu Mar 17 16:36:06 PDT 2022
```

**Comment by jmishrarh** [comment]
I am also facing the same problem while deploying ramen.
Tue Mar 22 18:16:46 IST 2022
+ until_true_or_n 90 kubectl --context hub -n open-cluster-management-hub
wait deployments --all --for condition=available --timeout 0
Tue Mar 22 18:16:46 IST 2022
error: no matching resources found
error: no matching resources found
error: no matching resources found
error: no matching resources found
error: no matching resources found
Could you please respond if there is a workaround (change of few lines ) to
fix in the script locally  and I can move ahead. I saw that there are a lot
of changes done in 408. Should we wait for 408 to merge?

Regards
Jolly

**Comment by hatfieldbrian** [comment]
> Could you please respond if there is a workaround (change of few lines ) to fix in the script locally and I can move ahead. I saw that there are a lot of changes done in 408. Should we wait for 408 to merge?

To fix this issue, the registration-operator image's tag needs to be changed from:
`quay.io/open-cluster-management/registration-operator:latest` to
`quay.io/open-cluster-management/registration-operator:0.1.0-c723e190c454110797d89e67bcb33250b35e1fd7`

The `registration_operator_kubectl` routine does this by producing two `kustomization.yaml` files, one for hub and one for spokes.

Hub:

```yaml
resources:
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/crds/0000_01_operator.open-cluster-management.io_clustermanagers.crd.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/rbac/cluster_role.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/rbac/cluster_role_binding.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/operator/namespace.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/operator/service_account.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/cluster-manager/config/operator/operator.yaml
patchesJson6902:
  - target:
      group: apps
      version: v1
      kind: Deployment
      name: cluster-manager
      namespace: open-cluster-management
    patch: |-
      - op: test
        path: /spec/replicas
        value: 3
      - op: replace
        path: /spec/replicas
        value: 1
      - op: test
        path: /spec/template/spec/containers/0/image
        value: quay.io/open-cluster-management/registration-operator:latest
      - op: replace
        path: /spec/template/spec/containers/0/image
        value: quay.io/open-cluster-management/registration-operator:0.1.0-c723e190c454110797d89e67bcb33250b35e1fd7
```
Spoke:
```yaml
resources:
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/crds/0000_00_operator.open-cluster-management.io_klusterlets.crd.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/rbac/cluster_role.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/rbac/cluster_role_binding.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/operator/namespace.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/operator/service_account.yaml
  - https://raw.githubusercontent.com/stolostron/registration-operator/c723e190c454110797d89e67bcb33250b35e1fd7/deploy/klusterlet/config/operator/operator.yaml
patchesJson6902:
  - target:
      group: apps
      version: v1
      kind: Deployment
      name: klusterlet
      namespace: open-cluster-management
    patch: |-
      - op: test
        path: /spec/replicas
        value: 3
      - op: replace
        path: /spec/replicas
        value: 1
      - op: test
        path: /spec/template/spec/containers/0/image
        value: quay.io/open-cluster-management/registration-operator:latest
      - op: replace
        path: /spec/template/spec/containers/0/image
        value: quay.io/open-cluster-management/registration-operator:0.1.0-c723e190c454110797d89e67bcb33250b35e1fd7
```
To apply one of the above kustomizations:
```sh
kubectl --context <cluster name of hub or spoke>  apply -k <directory pathname of kustomization.yaml file>
```

By default there is one hub cluster named `hub` and two spoke clusters named `hub` and `cluster1`.

You could do this manually, but I recommend instead using the `ocm-minikube.sh` file contained in this pull request.  It may not be merged, but it has been tested.

**Comment by hatfieldbrian** [comment]
I found that deleting the `cluster-manager` deployment was needed for the different container image to be deployed:
> I reproduced this error using `latest` registration operator image, deleted the `cluster-manager` deployment, then redeployed it using the image matching the manifests and verified that the `open-cluster-management-hub deployments` were all `available`

I did this by issuing the following command:
```sh
kubectl --context hub -nopen-cluster-management delete deploy/cluster-manager
```
For example:
```sh
$ kubectl --context hub -nopen-cluster-management delete deploy/cluster-manager
deployment.apps "cluster-manager" deleted
```
@jmishrarh if you can still reproduce this issue feel free to report here any issues or success with fixing it

**Comment by hatfieldbrian** [comment]
@jmishrarh

1. Based on the failure point, the registration-operator deployment that was previously unavailable became available meaning this issue tracked here has been fixed.
2. The issue encountered is that the hub cluster, which is also a managed cluster, waited 1 minute for its managed cluster subscription deployment, `multicluster-operators-subscription-mc`, to become available and it never did.  Have you tried running the script again or even just the command that failed to see if it just needed to wait longer?  E.g.:
```sh
$ kubectl --context hub -nmulticluster-operators wait --for condition=available --timeout 0 deploy --all
deployment.apps/multicluster-operators-application condition met
deployment.apps/multicluster-operators-subscription condition met
deployment.apps/multicluster-operators-subscription-mc condition met
```

**Comment by jmishrarh** [comment]
The above command worked today.  I am facing another issue. Initially I got an error that osd0-cluster1 already exist. So I manually deleted the image. After that when I ran the script I found this error. 

+ virsh vol-info --pool minikube osd0-cluster1
error: failed to get vol 'osd0-cluster1'
error: Storage volume not found: no storage vol with matching path 'osd0-cluster1'

+ virsh vol-create-as --pool minikube --name osd0-cluster1 --capacity 32G --format qcow2
Vol osd0-cluster1 created

+ virsh domblkinfo --domain cluster1 --device vdb
error: failed to get domain 'cluster1'

+ sudo virsh attach-disk --domain cluster1 --source /home/jmishra-vm/.minikube/images/osd0-cluster1 --target vdb --persistent --driver qemu --subdriver qcow2 --targetbus virtio
error: Failed to attach disk
error: Requested operation is not valid: target vdb already exists
